### DISCUSSION TITLE
FFmpeg raw is odd in rolling-release-2.2.x

### DISCUSSION THREAD

**Discussion by Warblefly** [savonet/liquidsoap#3308] [User Support]
**Describe the bug**

After upgrading to rolling-release-2.2.x, passing a source to an FFmpeg filter produces a type error. The error persists when the ocaml-ffmpeg lib is upgraded to the unreleased 1.1.9 (from the 1.1.8 available through opam)

```
At ./test.liq, line 43, char 20:
  a = audio_process(a)

Error 5: this value has type
  source(_) (inferred at ffmpeg.liq, line 19, char 2-66)
but it should be a subtype of the type of the value at ./test.liq, line 29, char 18
  ffmpeg.audio.raw(_)

```
**To Reproduce**

The filter is wrapped by a function to convert Liquidsoap's internal audio to FFmpeg raw audio:

```
def audio_process_wrap(s) =
  a = ffmpeg.raw.encode.audio(%ffmpeg(%audio.raw), s)
  a = audio_process(a)   #[this is the line 43, referred to above]
  ffmpeg.raw.decode.audio(a)
end 
```

But the filter at its first line throws the Error 5 about the type

```
def audio_process(s) =     # [This is the line 29 referred to above]
  def mkfilter(graph) =
    s = ffmpeg.filter.audio.input(graph, s)
.... etc.
```

**Expected behavior**

The filter should work normally, as in version 2.1.x

**Version details**

OS: Debian
Version: testing

**Install method**
Via opam and, for later git versions, via git, built with dune (no errors)

**Comment by toots** [comment]
Hi @Warblefly ! 

The filters operate directly at the track level in `2.2.0`. This should be a pretty simple lift though:

```liquidsoap
def audio_process(s) = 
  def mkfilter(graph) =
    let {audio} = source.tracks(s)
    audio = ffmpeg.filter.audio.input(graph, audio)
...
    source({
      audio = audio,
      metadata = track.metadata(audio),
      track_marks = track.track_marks(audio)
    })
```
